src/journal-remote/journal-remote-main.c

```diff
@@ -212,8 +239,19 @@ static int process_http_upload(
         if (*upload_data_size) {
                 log_trace("Received %zu bytes", *upload_data_size);
 
-                r = journal_importer_push_data(&source->importer,
-                                               upload_data, *upload_data_size);
+                if (source->compression != COMPRESSION_NONE) {
+                        _cleanup_free_ char *buf = NULL;
+                        size_t buf_size;
+
+                        r = decompress_blob(source->compression, upload_data, *upload_data_size, (void **) &buf, &buf_size, 0);
+                        if (r < 0) {
+                                log_warning_errno(r, "Decompression of received blob failed, %p", connection);
+                                return MHD_NO;
```

Comment:
Not sure, but maybe `mhd_resopond()` or `mhd_respondf()` is better?
See error handling in `journal_importer_bytes_remaining()` below for example.